File: holoken-core/src/main/java/com/holokenmod/game/Game.java
package com.holokenmod.game;

import com.holokenmod.grid.Grid;
import com.holokenmod.grid.GridCell;
import com.holokenmod.grid.GridView;
import com.holokenmod.undo.UndoManager;

import java.util.List;

public class Game {
	private Grid grid;
	private GridView gridUI;
	private final UndoManager undoManager;
	private GameSolvedListener solvedListener = null;
	
	public Game(UndoManager undoManager) {
		this.undoManager = undoManager;
	}
	
	public Grid getGrid() {
		return this.grid;
	}
	
	public void setGridUI(GridView gridUI) {
		this.gridUI = gridUI;
	}
	
	public GridView getGridUI() {
		return gridUI;
	}
	
	public synchronized void enterNumber(final int number, final boolean removePossibles) {
		final GridCell selectedCell = grid.getSelectedCell();
		if (!grid.isActive()) {
			return;
		}
		if (selectedCell == null) {
			return;
		}
		clearLastModified();
		
		undoManager.saveUndo(selectedCell, false);
		
		selectedCell.setUserValue(number);
		if (removePossibles) {
			removePossibles(selectedCell);
		}
		
		if (grid.isActive() && grid.isSolved()) {
			if (grid.getSelectedCell() != null) {
				grid.getSelectedCell().setSelected(false);
				grid.getSelectedCell().getCage().setSelected(false);
			}
			
			grid.setActive(false);
			
			if (this.solvedListener != null) {
				this.solvedListener.puzzleSolved();
			}
		}
		
		gridUI.requestFocus();
		gridUI.invalidate();
	}
	
	public void setSolvedHandler(final GameSolvedListener listener) {
		this.solvedListener = listener;
	}
	
	public synchronized void enterPossibleNumber(final int number) {
		final GridCell selectedCell = grid.getSelectedCell();
		if (!grid.isActive()) {
			return;
		}
		if (selectedCell == null) {
			return;
		}
		clearLastModified();
		
		undoManager.saveUndo(selectedCell, false);
		
		if (selectedCell.isUserValueSet()) {
			final int oldValue = selectedCell.getUserValue();
			selectedCell.clearUserValue();
			selectedCell.togglePossible(oldValue);
		}
		
		selectedCell.togglePossible(number);
		
		gridUI.requestFocus();
		gridUI.invalidate();
	}
	
	public void removePossibles(final GridCell selectedCell) {
		final List<GridCell> possibleCells =
				grid.getPossiblesInRowCol(selectedCell);
		for (final GridCell cell : possibleCells) {
			undoManager.saveUndo(cell, true);
			cell.setLastModified(true);
			cell.removePossible(selectedCell.getUserValue());
		}
	}
	
	public void setGrid(Grid grid) {
		this.grid = grid;
	}
	
	public void selectCell() {
		final GridCell selectedCell = grid.getSelectedCell();
		if (!grid.isActive()) {
			return;
		}
		if (selectedCell == null) {
			return;
		}
		
		gridUI.requestFocus();
		gridUI.invalidate();
	}
	
	public void eraseSelectedCell() {
		final GridCell selectedCell = grid.getSelectedCell();
		
		if (!grid.isActive()) {
			return;
		}
		if (selectedCell == null) {
			return;
		}
		
		if (selectedCell.isUserValueSet() || selectedCell.getPossibles().size() > 0) {
			clearLastModified();
			undoManager.saveUndo(selectedCell, false);
			selectedCell.clearUserValue();
		}
	}
	
	public boolean setSinglePossibleOnSelectedCell(boolean rmpencil) {
		final GridCell selectedCell = grid.getSelectedCell();
		if (!grid.isActive()) {
			return false;
		}
		if (selectedCell == null) {
			return false;
		}
		
		if (selectedCell.getPossibles().size() == 1) {
			clearLastModified();
			undoManager.saveUndo(selectedCell, false);
			selectedCell.setUserValue(selectedCell.getPossibles().iterator().next());
			
			if (rmpencil) {
				removePossibles(selectedCell);
			}
		}
		
		gridUI.requestFocus();
		gridUI.invalidate();
		
		return true;
	}
	
	public void clearUserValues() {
		grid.clearUserValues();
		
		gridUI.invalidate();
	}
	
	public void clearLastModified() {
		grid.clearLastModified();
		
		gridUI.invalidate();
	}
	
	public boolean solveSelectedCage() {
		final GridCell selected = grid.getSelectedCell();
		
		if (selected == null) {
			return false;
		}
		
		grid.solveSelectedCage();
		
		gridUI.invalidate();
		
		return true;
	}
	
	public void solveGrid() {
		grid.solveGrid();
		
		gridUI.invalidate();
	}
	
	public void markInvalidChoices() {
		grid.markInvalidChoices();
		
		gridUI.invalidate();
	}
	
	public boolean revealSelectedCell() {
		final GridCell selectedCell = grid.getSelectedCell();
		
		if (selectedCell == null) {
			return false;
		}
		
		selectedCell.setUserValue(selectedCell.getValue());
		selectedCell.setCheated(true);
		
		gridUI.invalidate();
		
		return true;
	}
	
	public void undoOneStep() {
		clearLastModified();
		undoManager.restoreUndo();
		gridUI.invalidate();
	}
	
	public void restartGame() {
		clearUserValues();
		grid.setActive(true);
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/game/Game.kt
package com.holokenmod.game

import com.holokenmod.grid.Grid
import com.holokenmod.grid.GridCell
import com.holokenmod.grid.GridSolveService
import com.holokenmod.grid.GridView
import com.holokenmod.undo.UndoManager

data class Game(
    var grid: Grid,
    var undoManager: UndoManager,
    var gridUI: GridView
) {
    private var solvedListener: GameSolvedListener? = null

    private val gridSolveService = GridSolveService(grid)

    private val gridCreationListeners = mutableListOf<GridCreationListener>()

    fun addGridCreationListener(gridCreationListener: GridCreationListener) {
        gridCreationListeners += gridCreationListener
    }

    fun updateGrid(newGrid: Grid) {
        grid = newGrid
        gridUI.grid = grid

        gridCreationListeners.forEach { it.freshGridWasCreated() }
    }

    @Synchronized
    fun enterNumber(number: Int, removePossibles: Boolean) {
        val selectedCell = grid.selectedCell
        if (!grid.isActive) {
            return
        }
        if (selectedCell == null) {
            return
        }
        clearLastModified()
        undoManager.saveUndo(selectedCell, false)
        selectedCell.setUserValueExtern(number)
        if (removePossibles) {
            removePossibles(selectedCell)
        }
        if (grid.isActive && grid.isSolved) {
            selectedCell.isSelected = false
            selectedCell.cage?.setSelected(false)
            grid.isActive = false
            solvedListener?.puzzleSolved()
        }

        grid.userValueChanged()

        gridUI.requestFocus()
        gridUI.invalidate()
    }

    fun setSolvedHandler(listener: GameSolvedListener?) {
        solvedListener = listener
    }

    @Synchronized
    fun enterPossibleNumber(number: Int) {
        val selectedCell = grid.selectedCell ?: return
        if (!grid.isActive) {
            return
        }
        clearLastModified()
        undoManager.saveUndo(selectedCell, false)
        if (selectedCell.isUserValueSet) {
            val oldValue = selectedCell.userValue
            selectedCell.clearUserValue()
            selectedCell.togglePossible(oldValue)
            grid.userValueChanged()
        }
        selectedCell.togglePossible(number)
        gridUI.requestFocus()
        gridUI.invalidate()
    }

    private fun removePossibles(selectedCell: GridCell) {
        val possibleCells = grid.getPossiblesInRowCol(selectedCell)
        for (cell in possibleCells) {
            undoManager.saveUndo(cell, true)
            cell.isLastModified = true
            cell.removePossible(selectedCell.userValue)
        }
    }

    fun selectCell(cell: GridCell) {
        grid.selectedCell = cell

        for (c in grid.cells) {
            c.isSelected = false
            c.cage?.setSelected(false)
        }

        cell.isSelected = true
        cell.cage?.setSelected(true)

        gridUI.requestFocus()
        gridUI.invalidate()
    }

    fun eraseSelectedCell() {
        val selectedCell = grid.selectedCell ?: return

        if (!grid.isActive) {
            return
        }

        if (selectedCell.isUserValueSet || selectedCell.possibles.isNotEmpty()) {
            clearLastModified()
            undoManager.saveUndo(selectedCell, false)
            selectedCell.clearUserValue()
            selectedCell.clearPossibles()
            grid.userValueChanged()
        }
    }

    fun setSinglePossibleOnSelectedCell(removePossibles: Boolean): Boolean {
        val selectedCell = grid.selectedCell ?: return false

        if (!grid.isActive) {
            return false
        }

        if (selectedCell.possibles.size == 1) {
            enterNumber(selectedCell.possibles.first(), removePossibles)
        }

        return true
    }

    private fun clearUserValues() {
        grid.clearUserValues()
        gridUI.invalidate()
    }

    private fun clearLastModified() {
        grid.clearLastModified()
        gridUI.invalidate()
    }

    fun solveSelectedCage(): Boolean {
        grid.selectedCell ?: return false
        gridSolveService.solveSelectedCage()
        gridUI.invalidate()
        return true
    }

    fun solveGrid() {
        gridSolveService.solveGrid()
        gridUI.invalidate()
    }

    fun revealSelectedCell(): Boolean {
        gridSolveService.revealSelectedCell()
        gridUI.invalidate()
        return true
    }

    fun markInvalidChoices(showDupedDigits: Boolean) {
        grid.markInvalidChoices(showDupedDigits)
        gridUI.invalidate()
    }

    fun undoOneStep() {
        clearLastModified()
        undoManager.restoreUndo()
        gridUI.invalidate()
    }

    fun restartGame() {
        clearUserValues()
        grid.isActive = true
    }

    fun restoreUndo() {
        undoManager.restoreUndo()
    }

    fun clearUndoList() {
        undoManager.clear()
    }

    fun undo() {
        clearLastModified()
        restoreUndo()
        gridUI.invalidate()
    }

    fun solveAllMissingCells() {
        grid.cells.forEach {
            selectCell(it)
            enterNumber(it.value, true)
        }
    }
}
